Add initial code

arcusfelis opened the pull request:
Check [readme](https://github.com/esl/cets/blob/rebased/README.md) for description

arcusfelis committed:
Add an empty application and a test suite

arcusfelis committed:
Add ETS replication logic (working simple test)

Use persistent term for metadata

Monitor a proxy process

arcusfelis committed:
Add logging and long task tracking

arcusfelis committed:
Lock joining

Report time to get the joining lock

arcusfelis committed:
Add auto-discovery support

arcusfelis committed:
Add handle_down

arcusfelis committed:
Add delete and multidelete

arcusfelis committed:
Allow to add tables to disco dynamically

Register name for disco

arcusfelis committed:
Use nosuspend when sending messages

arcusfelis committed:
Use run_safely

Remove try-catch

arcusfelis committed:
Support discovery backends

arcusfelis committed:
Use OTP logger

arcusfelis committed:
Remove remote_just_add_node_to_schema function

arcusfelis committed:
Split handle_join

arcusfelis committed:
Route deletes/inserts though the local server

arcusfelis committed:
Remove kiss_pt

arcusfelis committed:
Properly pause to avoid race conditions

arcusfelis committed:
Move joining logic into a separate file

arcusfelis committed:
Dedup code with short_call

arcusfelis committed:
Merge insert_to_remote_nodes and delete_from_remote_nodes

arcusfelis committed:
Reorder clauses in handle_call

arcusfelis committed:
Fix stop output in kiss_proxy

arcusfelis committed:
Reorder functions for better readability

arcusfelis committed:
Add events_are_applied_in_the_correct_order_after_unpause test case

Add request versions of insert/delete functions

arcusfelis committed:
Use monitor table

arcusfelis committed:
Add MonTab cleaner

arcusfelis committed:
Send remote_down from the local server

arcusfelis committed:
Add description into the file header

arcusfelis committed:
Simplify replicate2

arcusfelis committed:
Add mon_cleaner_works testcase

arcusfelis committed:
Add info function

arcusfelis committed:
Allow to pass atom to sync/1

arcusfelis committed:
Log unexpected messages

arcusfelis committed:
Add types

arcusfelis committed:
Set message_queue_data=off_heap

arcusfelis committed:
Optimize copying during join

arcusfelis committed:
Do joining in a new process

arcusfelis committed:
Put try..catch logic into kiss_safety

arcusfelis committed:
Rename kiss to cets

arcusfelis committed:
Add insert_many_request and insert_many API functions

arcusfelis committed:
Use erlang:system_time api

arcusfelis committed:
Remove cets_safety module

arcusfelis committed:
Avoid deadlocks in sync/1

arcusfelis committed:
Add Github CI

Update readme

arcusfelis committed:
Allow multiple pauses

arcusfelis committed:
Add servers_to_pids and has_remote_pid helper functions

arcusfelis committed:
Add dialyzer task for CI

arcusfelis committed:
Separate operations for insert and insert_many

Separate operations for delete and delete_many
Add more types

arcusfelis committed:
Add unpause_twice and pause_multiple_times testcases

arcusfelis committed:
Use long_call function

chrzaszcz commented on README.md:
```diff
@@ -0,0 +1,55 @@
+- `delete(Server, Key)` - deletes an object from the table.
+- `delete_many(Server, Keys)` - deletes several objects from the table.
+
+cets_join module contains the merging logic.
```
```suggestion
`cets_join` module contains the merging logic.
```

chrzaszcz commented on README.md:
```diff
@@ -0,0 +1,55 @@
+
+cets_join module contains the merging logic.
+
+cets_discovery module handles search of new nodes.
```
```suggestion
`cets_discovery` module handles search of new nodes.
```
...and it would be good to format the remainder of this page as well.

arcusfelis committed:
Move call specific code into cets_call

To split the business logic from the helpers logic

arcusfelis committed:
Update file docs

arcusfelis committed:
Fix dialyzer

pawlooss1 commented on src/cets_long.erl:
```diff
@@ -0,0 +1,73 @@
+        after
+            Diff = diff(Start),
+            ?LOG_INFO(Info#{what => long_task_finished, time_ms => Diff}),
+            Pid ! stop
```
This indent is unnecessary big. I see a similar case in some other places as well. What do you think about plugging in some code formatter, such as https://github.com/WhatsApp/erlfmt? I think it could help us keep the formatting consistent.

arcusfelis replied:
I have so many questions to erlfmt :) 
At least for nodejs formatters you can configure the style.

That aligning when `when` is used is kinda wtf:

```erlang

resolve(#amp_strategy{deliver = [Value | _]}, #amp_rule{
    condition = deliver, value = Value, action = Action
}) when
    Action == drop orelse Action == error
->
    match;
resolve(#amp_strategy{'match-resource' = Value}, #amp_rule{
    condition = 'match-resource', value = any
}) when
    Value /= undefined
->
    match;
resolve(#amp_strategy{'match-resource' = Value}, #amp_rule{
    condition = 'match-resource', value = Value
}) ->
```

And brackets:

```erlang
    CommandTags = ets:match(
        ejabberd_commands,
        #ejabberd_commands{
            name = '$1',
            tags = '$2',
            _ = '_'
        }
    ),
```

Maybe we should use it in mongooseim ;)

pawlooss1 commented on src/cets_long.erl:
```diff
@@ -0,0 +1,73 @@
+    Pid = spawn_mon(Info, Parent, Start),
+    try
+            case Catch of
+                true -> just_run_safely(Info#{what => long_task_failed}, Fun);
```
Putting `what => long_task_failed` in this line suggests that our task has already failed. I think it would be more straightforward if we just put it directly in the logging in line 71.

pawlooss1 commented on src/cets_call.erl:
```diff
@@ -0,0 +1,109 @@
+-type server_ref() :: cets:server_ref().
+-type long_msg() :: cets:long_msg().
+
+%% Do gen_server:call with better error reporting.
```
It should be `Does` instead of `Do`.

pawlooss1 commented on src/cets_call.erl:
```diff
@@ -0,0 +1,109 @@
+-type long_msg() :: cets:long_msg().
+
+%% Do gen_server:call with better error reporting.
+%% It would print a warning if the call takes too long.
```
I'd change "It would print" to "It will log".

pawlooss1 commented on src/cets_call.erl:
```diff
@@ -0,0 +1,109 @@
+%% remote_down is sent by the local server, if the remote server is down.
+wait_for_updated(Mon, {Servers, MonTab}) ->
+    try
+        wait_for_updated2(Mon, Servers)
```
In MongooseIM I saw a naming convention for such case `do_FunctionName` and I like it. Maybe we could rename this one to follow it? I think that giving number suffixes to function names can be confusing for the reader.

pawlooss1 commented on rebar.config:
```diff
@@ -0,0 +1,9 @@
+     no_opaque, no_fail_call, no_contracts, no_behaviours,
+     no_undefined_callbacks, unmatched_returns, error_handling,
+     underspecs
+     % overspecs, specdiffs
```
Please either remove or uncomment this line.

pawlooss1 commented on src/cets.app.src:
```diff
@@ -0,0 +1,9 @@
+  {registered, []},
+  {applications, [kernel, stdlib]},
+  {env, []}
+% {mod, {cets_app, []}
```
Please either remove or uncomment this line.

pawlooss1 commented on src/cets_discovery_file.erl:
```diff
@@ -0,0 +1,28 @@
+init(Opts) ->
+    Opts.
+
+get_nodes(State = #{disco_file := Filename}) ->
```
I think a `spec` here would be useful for further development. Especially that it is a callback for the `cets_discovery` behaviour.

pawlooss1 commented on src/cets_join.erl:
```diff
@@ -0,0 +1,78 @@
+    Info2 = Info#{local_pid => LocalPid,
+                  remote_pid => RemotePid, remote_node => node(RemotePid)},
+    F = fun() -> join1(LockKey, Info2, LocalPid, RemotePid) end,
+    cets_long:run_safely(Info2#{what => join_failed}, F).
```
If I get this correctly, `what` key of this map is going to be overwritten in cets_long.erl in every logging line. Could you check it?

chrzaszcz replied:
> If I get this correctly, `what` key of this map is going to be overwritten in cets_long.erl in every logging line. Could you check it?

What's more, this can be misleading especially when debugging/tracing. Finding `what => join_failed` in the arguments would suggest that something has already failed.

chrzaszcz replied:
Additionally, there will be more `cets_long` calls inside this `cets_long` call. I don't think we need doubled monitors. Maybe skip the outer `cets_long` entirely and just catch errors?

chrzaszcz commented on src/cets.erl:
```diff
@@ -0,0 +1,420 @@
+stop(Server) ->
+    gen_server:stop(Server).
+
+-spec dump(server_ref()) -> Records :: [tuple()].
```
```suggestion
-spec dump(table_name()) -> Records :: [tuple()].
```

pawlooss1 commented on test/cets_SUITE.erl:
```diff
@@ -0,0 +1,259 @@
+end_per_testcase(_, _Config) ->
+    ok.
+ 
+test_multinode(Config) ->
```
This test checks a few thing, e.g. joining nodes (both before and after adding data), inserting values, deleting values, bulk operations. Could you split it to minimal test cases where each one tests exactly one thing?

arcusfelis replied:
I added more tests for each operation

chrzaszcz replied:
Ok, but it just doesn't help with the readability of this particular one.

pawlooss1 reviewed (commented):
Nice work 👍 I left some comments that I think could improve readability and maintainability.

chrzaszcz commented on run_test.sh:
```diff
@@ -0,0 +1,2 @@
+#!/usr/bin/env bash
+./rebar3 ct --sname=ct1
```
There is no `rebar3` in this repo. I would just do:

```suggestion
rebar3 ct --sname=ct1
```

chrzaszcz commented on src/cets.erl:
```diff
@@ -0,0 +1,420 @@
+
+%% Table and server has the same name
+%% Opts:
+%% - handle_down = fun(#{remote_pid => Pid, table => Tab})
```
```suggestion
%% - handle_down = fun(#{remote_pid := Pid, table := Tab})
```

chrzaszcz commented on src/cets_join.erl:
```diff
@@ -0,0 +1,78 @@
+        cets:sync(RemotePid),
+        {ok, LocalDump} = remote_or_local_dump(LocalPid),
+        {ok, RemoteDump} = remote_or_local_dump(RemotePid),
+        RemF = fun(Pid) -> cets:send_dump(Pid, LocPids, LocalDump) end,
```
Whole table sent in one go? It might be ok for now, but I am thinking about the case where we need to send the entire session table from one node to another when it joins the cluster. We would need to allocate a lot of memory on the process heap for this operation. Could you check what happens for a large table?

arcusfelis replied:
There would be a spike of memory usage (probably we need 2 times of the ETS table size).
For the real life ejabberd_sm tables - would be sent pretty fast.

Though, it is the fastest way to send. Also, could cause distributed connection processes to stuck for a second. 

I was thinking if we need pagination to be added - it could be done separately (separate PR), because it would be easier to review the main PR.

chrzaszcz commented on src/cets_join.erl:
```diff
@@ -0,0 +1,78 @@
+%% Adds a node to a cluster.
+%% Writes from other nodes would wait for join completion.
+%% LockKey should be the same on all nodes.
+join(LockKey, Info, LocalPid, RemotePid) when is_pid(LocalPid), is_pid(RemotePid) ->
```
Please add type specs

Also prevent cets from killing the processes (and losing ETS data) in the event of accidentally calling it with LocalPid same as RemotePid. I checked it and it crashes.

chrzaszcz commented on src/cets.erl:
```diff
@@ -0,0 +1,420 @@
+            add_servers2(OtherPids, Servers)
+    end;
+add_servers2([], _Servers) ->
+    [].
```
Minor, style: I think it would be a good idea to have a separate function for adding a single server.
```suggestion
add_servers(Pids, Servers) ->
    lists:sort(lists:foldl(fun(Pid) -> add_server(Pid, Servers) end, Servers, Pids)).

add_server(RemotePid, Servers)
  when is_pid(RemotePid), RemotePid =/= self() ->
    case has_remote_pid(RemotePid, Servers) of
        false ->
            erlang:monitor(process, RemotePid),
            [RemotePid | Servers];
        true ->
            ?LOG_INFO(#{what => already_added,
                        remote_pid => RemotePid, remote_node => node(RemotePid)}),
            Servers
    end.
```

chrzaszcz commented on src/cets.erl:
```diff
@@ -0,0 +1,420 @@
+init({Tab, Opts}) ->
+    process_flag(message_queue_data, off_heap),
+    MonTab = list_to_atom(atom_to_list(Tab) ++ "_mon"),
+    _ = ets:new(Tab, [ordered_set, named_table, public]),
+    _ = ets:new(MonTab, [public, named_table]),
```
This pattern matching does nothing.

arcusfelis replied:
it prevents from dialyzer warning
```
src/cets.erl
Line 241 Column 5: Expression produces a value of type atom() | ets:tid(), but this value is unmatched
Line 242 Column 5: Expression produces a value of type atom() | ets:tid(), but this value is unmatched
===> Warnings written to _build/default/25.0.1.dialyzer_warnings
```

chrzaszcz commented on src/cets_mon_cleaner.erl:
```diff
@@ -0,0 +1,78 @@
+
+-spec handle_check(state()) -> state().
+handle_check(State = #{mon_tab := MonTab}) ->
+    check_loop(ets:tab2list(MonTab), MonTab),
```
After you call `cets:stop/1`, the cleaner will crash here with `badarg`.

arcusfelis replied:
added gen_server:stop into terminate of the main server.

chrzaszcz commented on src/cets_join.erl:
```diff
@@ -0,0 +1,78 @@
+%% LockKey should be the same on all nodes.
+join(LockKey, Info, LocalPid, RemotePid) when is_pid(LocalPid), is_pid(RemotePid) ->
+    Info2 = Info#{local_pid => LocalPid,
+                  remote_pid => RemotePid, remote_node => node(RemotePid)},
```
Is there a point in calling them `local` and `remote` if even in tests the first Pid is the remote one? Either ensure the first one is local, or rename them.

chrzaszcz commented on src/cets_long.erl:
```diff
@@ -0,0 +1,73 @@
+%% This allows to reduce GC on the parent process
+%% Wait for function to finish
+%% Handles errors
+run_spawn(Info, F) ->
```
Please add specs for exported functions.

chrzaszcz commented on src/cets_join.erl:
```diff
@@ -0,0 +1,78 @@
+    try
+        cets:sync(LocalPid),
+        cets:sync(RemotePid),
+        {ok, LocalDump} = remote_or_local_dump(LocalPid),
+        {ok, RemoteDump} = remote_or_local_dump(RemotePid),
```
Here the local vs remote semantics gets particularly blurry. I mean: the Local one can be a remote one instead, it is so in the tests.

arcusfelis replied:
Yes, in tests both pids could be local.
But once discovery module calls `join`, Local and Remote pids would be like that.
I could call it Server1 and Server2, but it would be harder to read (probably).

But basically, there is no difference between these two servers, exchange is bi-directional. They both send dumps to each other.

arcusfelis replied:
it is primarily for _logging_, otherwise we would have something like `node1` and `node2`, `pid1` and `pid2`.

chrzaszcz replied:
Actually in some tests the local one is the remote one, they are **swapped**, that's what I meant.

chrzaszcz commented on src/cets_mon_cleaner.erl:
```diff
@@ -0,0 +1,78 @@
+
+-spec schedule_check(state()) -> state().
+schedule_check(State = #{interval := Interval, timer_ref := OldRef}) ->
+    _ = erlang:cancel_timer(OldRef),
```
What is the point of pattern matching to anything?

arcusfelis replied:
for dialyzer, added as a comment

chrzaszcz commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,50 @@
+    runs-on: ubuntu-latest
+
+    container:
+      image: erlang:23
```
Update Erlang version.

chrzaszcz commented on .gitignore:
```diff
@@ -0,0 +1,4 @@
+_build/
+rebar3
```
There is no `rebar3` in the repo.

chrzaszcz commented on src/cets.erl:
```diff
@@ -0,0 +1,420 @@
+replicate(From, Msg, #{mon_tab := MonTab, other_servers := Servers}) ->
+    %% Reply would be routed directly to FromPid
+    Msg2 = {remote_op, From, Msg},
+    replicate2(Servers, Msg2),
```
I don't think we gain anything by the increased amount of code and the additional function with a `2` suffix.  `lists:foreach` is also fine for me - easier to read, debug, understand...

```suggestion
    [send_to_remote(RemotePid, Msg2) || RemotePid <- Servers],
```

chrzaszcz commented on src/cets.erl:
```diff
@@ -0,0 +1,420 @@
+    ok.
+
+apply_backlog(State = #{backlog := Backlog}) ->
+    apply_backlog_ops(lists:reverse(Backlog), State),
```
Same as above.

chrzaszcz commented on src/cets_discovery.erl:
```diff
@@ -0,0 +1,146 @@
+-callback init(map()) -> backend_state().
+-callback get_nodes(backend_state()) -> {get_nodes_result(), backend_state()}.
+
+start(Opts) ->
```
Please add specs to the API functions. `Opts` need a spec.

chrzaszcz commented on src/cets_discovery.erl:
```diff
@@ -0,0 +1,146 @@
+    State#{timer_ref := TimerRef}.
+
+cancel_old_timer(#{timer_ref := OldRef}) when is_reference(OldRef) ->
+    _ = erlang:cancel_timer(OldRef),
```
What is the point of this pattern match? It has no effect.

arcusfelis replied:
prevents from dialyzer warning.

chrzaszcz replied:
Which option triggers this warning? I see a lot of them enabled. Sometimes it might be better to disable a check if it forces us to write unnecessary code.

chrzaszcz reviewed (commented):
It looks promising 🙂 I added some comments.

chrzaszcz commented on src/cets.erl:
```diff
@@ -0,0 +1,420 @@
+init({Tab, Opts}) ->
+    process_flag(message_queue_data, off_heap),
+    MonTab = list_to_atom(atom_to_list(Tab) ++ "_mon"),
+    _ = ets:new(Tab, [ordered_set, named_table, public]),
```
Another note: why `ordered_set`? It slows down read time. Also: how about allowing some customisation instead of hardcoding the table type?

arcusfelis replied:
ordered_set is required for ejabberd_sm to work - for reads all sessions of the user, for example.

arcusfelis committed:
Autoformat with erlfmt

arcusfelis committed:
Update erlang version to 25

arcusfelis committed:
Add specs for cets_discovery

Address some review comments

arcusfelis committed:
Add more granular tests

Test each operation of test_multinode separately

arcusfelis committed:
Remove unnecessary cets_long:run call in join

arcusfelis committed:
Add comment about matching result for dialyzer

arcusfelis committed:
Add specs for cets_long / cets_join

arcusfelis committed:
Handle join_with_the_same_pid

prevent cets from killing the processes (and losing ETS data) in the event of accidentally calling it with LocalPid same as RemotePid

arcusfelis committed:
Stop monitor process properly

Expose mon_pid in cets:info

arcusfelis committed:
Remove unused cets_long:run/2

chrzaszcz commented on test/cets_SUITE.erl:
```diff
@@ -0,0 +1,359 @@
+    {ok, Pid2} = start(Node2, Tab),
+    {ok, Pid3} = start(Node3, Tab),
+    {ok, Pid4} = start(Node4, Tab),
+    ok = join(Node1, Tab, Pid3, Pid1),
```
This happens a few times: the first pid is remote, and the second is local, while in the `join` function it is the other way around. Maybe don't swap the order to make it easier to understand?

chrzaszcz reviewed (approved):
Looks good :+1:

arcusfelis committed:
Fix the order of arguments in join in tests
